src/docs/sphinx/dev_guide/tensor_dual.rst

```diff
@@ -0,0 +1,405 @@
+track the gradients of intermediate values is more expensive than
+manually writing code for the derivatives.
+
+However, by supporting both manually-written derivatives and ``dual``
+numbers, users can choose to calculate derivatives in whatever manner is
+appropriate for their problem: manually-written gradients for
+performance-critical codepaths, and automatic differentiation for
+iterating quickly on prototypes and research.
+
+Some additional resources on the theory and implementation of automatic differentiation
+are given below:
+
+`Slides on AD Theory <https://www.cs.toronto.edu/~rgrosse/courses/csc321_2018/slides/lec10.pdf>`_
+
+`Article demonstrating how AD applies to a computational graph <https://towardsdatascience.com/automatic-differentiation-explained-b4ba8e60c2ad>`_
+
+`C++ tools and libraries for AD <http://www.autodiff.org/?module=Tools&language=C%2FC%2B%2B>`_
+
+.. _header-n276:
+
+Using ``tensor`` and ``dual`` together
+--------------------------------------
+
+In the previous example, :math:`f` was a function with a scalar input
+and scalar output. In practice, most of the functions we care about are
+more interesting. For example, an isotropic linear elastic material in
+solid mechanics has the following stress-strain relationship:
+
+.. math:: \sigma = \lambda \; \text{tr}(\epsilon) \; \mathbf{I} + 2 \; \mu \; \epsilon
+
+or, in C++:
+
+.. code-block:: cpp
+
+   double lambda = 2.0;
+   double mu = 1.0;
+   static constexpr auto I = Identity<3>();
+   auto stress = [=](auto strain){ return lambda * tr(strain) * I + 2 * mu * strain; };
+
+That is, ``stress()`` takes a ``tensor<double,3,3>`` as input, and
+outputs a ``tensor<double, 3, 3>``:
+
+.. code-block:: cpp
+
+   tensor< double, 3, 3 > epsilon = {...};
+   tensor< double, 3, 3 > sigma = stress(epsilon);
+
+In general, each part of a function's output can depend on each part of
+its inputs. So, in this example the gradient could potentially have up
+to 81 components:
+
+.. math:: \frac{\partial \sigma_{ij}}{\partial \epsilon_{kl}}, \qquad i,j,k,l \in {1,2,3}
+
+If we promote the input argument to a tensor of dual numbers, we can
+compute these derivatives automatically:
+
+.. code-block:: cpp
+
+   tensor< double, 3, 3 > epsilon = {...};
+   tensor< dual< tensor< double, 3, 3 > >, 3, 3 > sigma = stress(make_dual(epsilon));
+
+Now, ``sigma`` contains value and gradient information that can be
+understood in the following way
+
+.. math:: \texttt{sigma[i][j].value} = \sigma_{ij} \qquad \texttt{sigma[i][j].gradient[k][l]} = \frac{\partial \sigma_{ij}}{\partial \epsilon_{kl}}
+
+There are also convenience routines to extract all the values and
+gradient terms into their own tensors of the appropriate shape:
+
+.. code-block:: cpp
+
+   // as before
+   tensor< dual< tensor< double, 3, 3 > >, 3, 3 > sigma = stress(make_dual(epsilon));
+
+   // extract the values
+   tensor< double, 3, 3 > sigma_values = get_value(sigma);
```

Comment:
Is there a reason these aren't just class getter's?

```
tensor< double, 3, 3 > sigma_values = sigma.value();
tensor< double, 3, 3, 3, 3 > sigma_gradients = sigma.gradient();
```

Reply:
Ignore this... @joshessman-llnl pointed out the bigger picture I hadn't gotten to yet in the review...